File: rsbullet/examples/pybullet_examples/franka_ik_target_pose.py
import math
import time

import pybullet as p
import pybullet_data


def main() -> None:
    client_id = p.connect(p.GUI)
    p.setAdditionalSearchPath(pybullet_data.getDataPath())
    p.setGravity(0.0, 0.0, -9.81)
    p.setTimeStep(1.0 / 240.0)

    p.loadURDF("plane.urdf")
    robot_id = p.loadURDF("franka_panda/panda.urdf", [0.0, 0.0, 0.0], useFixedBase=True)

    arm_joint_indices = []
    for joint_index in range(p.getNumJoints(robot_id)):
        info = p.getJointInfo(robot_id, joint_index)
        if info[2] == p.JOINT_REVOLUTE:
            arm_joint_indices.append(joint_index)
            if len(arm_joint_indices) == 7:
                break

    if len(arm_joint_indices) != 7:
        raise RuntimeError("failed to find 7 Franka arm revolute joints")

    default_arm_pose = [0.0, -0.6, 0.0, -2.1, 0.0, 1.6, 0.8]
    for i, joint_index in enumerate(arm_joint_indices):
        p.resetJointState(robot_id, joint_index, default_arm_pose[i])

    end_effector_link_index = 11
    target_pos = [0.45, 0.00, 0.45]
    target_orn = p.getQuaternionFromEuler([math.pi, 0.0, 0.0])

    lower_limits = [
        -2.8973,
        -1.7628,
        -2.8973,
        -3.0718,
        -2.8973,
        -0.0175,
        -2.8973,
        0.0,
        0.0,
    ]
    upper_limits = [2.8973, 1.7628, 2.8973, -0.0698, 2.8973, 3.7525, 2.8973, 0.04, 0.04]
    joint_ranges = [u - l for l, u in zip(lower_limits, upper_limits)]
    rest_poses = default_arm_pose + [0.02, 0.02]

    tick = 0
    try:
        while p.isConnected(client_id):
            joint_poses = p.calculateInverseKinematics(
                robot_id,
                end_effector_link_index,
                target_pos,
                target_orn,
                lowerLimits=lower_limits,
                upperLimits=upper_limits,
                jointRanges=joint_ranges,
                restPoses=rest_poses,
                maxNumIterations=200,
                residualThreshold=1e-4,
            )

            for i, joint_index in enumerate(arm_joint_indices):
                p.setJointMotorControl2(
                    bodyUniqueId=robot_id,
                    jointIndex=joint_index,
                    controlMode=p.POSITION_CONTROL,
                    targetPosition=joint_poses[i],
                    force=200.0,
                )

            p.stepSimulation()
            time.sleep(1.0 / 240.0)

            tick += 1
            if tick % 240 == 0:
                ee_state = p.getLinkState(
                    robot_id,
                    end_effector_link_index,
                    computeLinkVelocity=0,
                    computeForwardKinematics=1,
                )
                ee_pos = ee_state[4]
                pos_err = math.sqrt(
                    (ee_pos[0] - target_pos[0]) ** 2
                    + (ee_pos[1] - target_pos[1]) ** 2
                    + (ee_pos[2] - target_pos[2]) ** 2
                )
                print(
                    "ee=({:.3f}, {:.3f}, {:.3f}), target=({:.3f}, {:.3f}, {:.3f}), pos_err={:.4f}".format(
                        ee_pos[0],
                        ee_pos[1],
                        ee_pos[2],
                        target_pos[0],
                        target_pos[1],
                        target_pos[2],
                        pos_err,
                    )
                )
    finally:
        if p.isConnected(client_id):
            p.disconnect(client_id)


if __name__ == "__main__":
    main()

File: rsbullet/examples/franka_ik_target_pose.rs
use std::{thread::sleep, time::Duration};

use nalgebra as na;
use rsbullet::{
    BulletResult, ControlModeArray, InverseKinematicsOptions, JointType, Mode, PhysicsClient,
    UrdfOptions,
};

fn main() -> BulletResult<()> {
    let mut client = PhysicsClient::connect(Mode::Gui)?;
    client
        .set_default_search_path()?
        .set_gravity([0., 0., -9.81])?
        .set_time_step(Duration::from_secs_f64(1.0 / 240.0))?;

    client.load_urdf("plane.urdf", None::<()>)?;

    let robot_id = client.load_urdf(
        "franka_panda/panda.urdf",
        Some(UrdfOptions {
            base: Some(na::Isometry3::translation(0.0, 0.0, 0.0)),
            use_fixed_base: true,
            ..Default::default()
        }),
    )?;

    let mut arm_joint_indices = Vec::new();
    for joint_index in 0..client.get_num_joints(robot_id) {
        let info = client.get_joint_info(robot_id, joint_index)?;
        if info.joint_type == JointType::Revolute {
            arm_joint_indices.push(joint_index);
            if arm_joint_indices.len() == 7 {
                break;
            }
        }
    }

    if arm_joint_indices.len() != 7 {
        return Err(rsbullet::BulletError::CommandFailed {
            message: "failed to find 7 Franka arm revolute joints",
            code: arm_joint_indices.len() as i32,
        });
    }

    let default_arm_pose = [0.0, -0.6, 0.0, -2.1, 0.0, 1.6, 0.8];
    for (i, joint_index) in arm_joint_indices.iter().enumerate() {
        client.reset_joint_state(robot_id, *joint_index, default_arm_pose[i], None)?;
    }

    let end_effector_link_index = 11;
    let target_pose = na::Isometry3::from_parts(
        na::Translation3::new(0.45, 0.00, 0.45),
        na::UnitQuaternion::from_euler_angles(std::f64::consts::PI, 0.0, 0.0),
    );

    let dof = client.compute_dof_count(robot_id) as usize;

    let mut tick = 0usize;
    while client.is_connected() {
        let current_arm_q: Vec<f64> = client
            .get_joint_states(robot_id, &arm_joint_indices)?
            .into_iter()
            .map(|s| s.position)
            .collect();

        let mut lower_full = vec![
            -2.8973, -1.7628, -2.8973, -3.0718, -2.8973, -0.0175, -2.8973,
        ];
        let mut upper_full = vec![2.8973, 1.7628, 2.8973, -0.0698, 2.8973, 3.7525, 2.8973];
        let mut rest_full = default_arm_pose.to_vec();

        while lower_full.len() < dof {
            lower_full.push(0.0);
            upper_full.push(0.04);
            rest_full.push(0.02);
        }
        let joint_ranges: Vec<f64> = lower_full
            .iter()
            .zip(upper_full.iter())
            .map(|(l, u)| u - l)
            .collect();

        let mut seed = current_arm_q;
        while seed.len() < dof {
            seed.push(0.02);
        }

        let ik_options = InverseKinematicsOptions {
            current_positions: Some(&seed),
            lower_limits: Some(&lower_full),
            upper_limits: Some(&upper_full),
            joint_ranges: Some(&joint_ranges),
            rest_poses: Some(&rest_full),
            max_iterations: Some(200),
            residual_threshold: Some(1e-4),
            ..Default::default()
        };

        let solution = client.calculate_inverse_kinematics(
            robot_id,
            end_effector_link_index,
            target_pose,
            &ik_options,
        )?;

        let target_positions: Vec<f64> = solution[..7].to_vec();
        client.set_joint_motor_control_array(
            robot_id,
            &arm_joint_indices,
            ControlModeArray::Position(&target_positions),
            None,
        )?;

        client.step_simulation()?;
        sleep(Duration::from_secs_f64(1.0 / 240.0));

        tick += 1;
        if tick.is_multiple_of(240) {
            let ee_state = client.get_link_state(robot_id, end_effector_link_index, true, false)?;
            let p = ee_state.world_link_frame.translation.vector;
            let t = target_pose.translation.vector;
            let err = (p - t).norm();
            println!(
                "ee=({:.3}, {:.3}, {:.3}), target=({:.3}, {:.3}, {:.3}), pos_err={:.4}",
                p.x, p.y, p.z, t.x, t.y, t.z, err
            );
        }
    }

    Ok(())
}
